traceevents/src/main/java/com/tomtom/kotlin/traceevents/Tracer.kt

```diff
@@ -674,6 +687,34 @@ public class Tracer private constructor(
             }
         }
 
+        /**
+         * Must be called after [traceEventChannel] has been updated, for example by calling
+         * [Channel.trySend] or removing elements by iterating over it.
+         *
+         * This updates whether the channel is empty. Though having to call this method to keep
+         * track of that is not ideal, there appears to be no better alternative. [Channel]s only
+         * allow receiving the elements within it (by a single collector) and do not allow observing
+         * updates to its queue.
+         */
+        @OptIn(ExperimentalCoroutinesApi::class)
+        private fun onTraceEventChannelUpdated() {
+            traceEventChannelIsEmpty.tryEmit(traceEventChannel.isEmpty)
```

Comment:
I'm not entirely happy with this approach but the only alternative I could find was to add a `while (true)` look in `waitForEmptyTraceEventsQueue` to wait for `traceEventChannel.isEmpty` which I liked less. I'm open to any suggestions.

Reply:
I don't think this solution really works when their is no event consumer at all. In that case `processTraceEvents` is never called and `traceEventChannelIsEmpty` will remain false forever.

The alternative I see is not to wait for it to be empty. You can send an event (not necessary a trace event) to the channel and than wait for it to comeback. Maybe easier is to have an `isTraceEventChannelEmpty` and poll it from the test (comparable to how espresso idling rules work)

Reply:
Indeed, if there's no consumer than the channel would not become empty. I'm actually thinking this seems like a problem in how the `Tracer` handles events when no consumer is registered regardless of this change. It leads to inconsistent behaviour where normally a consumer only gets events processed after it's been added, but now it can get events from before as well. Even worse: it depends on whether unrelated components add a consumer. Adding two consumers during startup closely after eachother means they'll get different events. If the goal of the current logic is to make sure trace events aren't lost, I don't think it's particularly effective because if you add a consumer that wants to consume everything from the start, it may still miss out on events that another consumer consumed.
I'm not sure if it's worth fixing or not.

I do like the goal of using whatever API we have in an `IdlingResource` in Android projects. With that goal we have a few options:
* Keep the solution as is but make `isTraceEventChannelEmpty` `public`. (Maybe with a better name.)
* Simplify the solution and simply expose an `isTraceEventChannelEmpty: Boolean` with custom `get()`.
* Use the trace event that you mentioned as callback.

Personally I think I'd opt for the second solution, with some documentation explaining that a consumer does have to be attached for the channel to become empty. Main reason for favouring solution 2 over 3 is that it seems like it'd be much simpler to implement in situations like `IdlingResource`s.

Do you agree?